src/Features/Core/Portable/Completion/Providers/AbstractEmbeddedLanguageCompletionProvider.cs

```diff
@@ -20,20 +22,36 @@ namespace Microsoft.CodeAnalysis.Completion.Providers
     /// Completions for an individual language are provided by
     /// <see cref="IEmbeddedLanguageFeatures.CompletionProvider"/>.
     /// </summary>
-    internal class EmbeddedLanguageCompletionProvider : CompletionProvider
+    internal abstract class AbstractEmbeddedLanguageCompletionProvider : CompletionProvider
     {
         public const string EmbeddedProviderName = "EmbeddedProvider";
 
-        private readonly ImmutableArray<IEmbeddedLanguage> _languageProviders;
+        private ImmutableArray<IEmbeddedLanguage> _languageProviders;
 
-        public EmbeddedLanguageCompletionProvider(IEmbeddedLanguagesProvider languagesProvider)
+        protected AbstractEmbeddedLanguageCompletionProvider()
         {
-            _languageProviders = languagesProvider?.Languages ?? ImmutableArray<IEmbeddedLanguage>.Empty;
+        }
+
+        protected ImmutableArray<IEmbeddedLanguage> GetLanguageProviders(Func<object, IEmbeddedLanguagesProvider> languagesProvider, object state)
+        {
+            if (_languageProviders.IsDefault)
+            {
+                ImmutableInterlocked.InterlockedInitialize(ref _languageProviders, languagesProvider(state)?.Languages ?? ImmutableArray<IEmbeddedLanguage>.Empty);
+            }
+
+            return _languageProviders;
         }
 
         public override bool ShouldTriggerCompletion(SourceText text, int caretPosition, CompletionTrigger trigger, OptionSet options)
         {
-            foreach (var language in _languageProviders)
+            Func<object, IEmbeddedLanguagesProvider> languagesProvider =
+                state =>
+                {
+                    var document = ((SourceText)state).GetOpenDocumentInCurrentContextWithChanges();
+                    return document.Project.LanguageServices.GetService<IEmbeddedLanguagesProvider>();
+                };
```

Comment:
why do these need to be Funcs that take objects?  that part isn't really clear to me.

Reply:
Fixed